### PR TITLE
Add multi layer decoder support for copynet seq2seq model

### PR DESCRIPTION
**Pull request by annajung**
Signed-off-by: Anna Jung (VMware) <antheaj@vmware.com>

Add support to multi-layer decoder in copynet seq2seq model.
Fixes https://github.com/allenai/allennlp/issues/4766

Joint work with @pab-vmware @divijbajaj @prajaktakini-vmware @agururajvais

### CODE REVIEW

File: tests/generation/models/copynet_test.py [hunk 1]
```diff
@@ -330,3 +338,9 @@ def setup_method(self):
 
     def test_model_can_train_save_load_predict(self):
         self.ensure_model_can_train_save_and_load(self.param_file, tolerance=1e-2)
+
+    def test_multi_layer_decoder_model_can_train_save_and_load(self):
+        param_overrides = json.dumps({"model": {"target_decoder_layers": 3}})
+        self.ensure_model_can_train_save_and_load(
+            self.param_file, tolerance=1e-2, overrides=param_overrides
+        )
```
Comment (epwalsh) [inline tests/generation/models/copynet_test.py:346]: I think we could do without this test. Not that it's a bad test, I just think we have everything covered already, and these tests have a lot of overhead.

**Review by epwalsh — COMMENTED** [review]
Thanks @annajung! This looks really great, I appreciate the attention to detail. I just have a couple comments 👇

File: allennlp_models/generation/models/copynet_seq2seq.py [hunk 2]
```diff
@@ -323,10 +348,22 @@ def _decoder_step(
...
-        )
+        # shape: (num_layers, group_size, decoder_output_dim)
+        decoder_hidden = state["decoder_hidden"]
+        # shape: (num_layers, group_size, decoder_output_dim)
+        decoder_context = state["decoder_context"]
+
+        if self._has_multiple_target_decoder_layers:
+            # shape: (group_size, source_sequence_length)
+            attentive_weights = self._attention(
+                decoder_hidden[0], state["encoder_outputs"], encoder_outputs_mask
```
[the remark this answers is not in the document]
Comment (annajung) [reply allennlp_models/generation/models/copynet_seq2seq.py]: I definitely wasn't sure about this one so was hoping to get some feedback here. Thanks for taking a look and makes sense to change it to -1! :)

**Comment by annajung** [comment]
@epwalsh PTAL when you can. Not totally sure why the GPU check is failing and there doesn't seem to be a way to retrigger the ci check without another commit?

File: allennlp_models/generation/models/copynet_seq2seq.py [hunk 3]
```diff
@@ -68,6 +68,8 @@ class CopyNetSeq2Seq(Model):
         in the batch and the second is a gold sequence for each item in the batch.
     initializer : `InitializerApplicator`, optional
         An initialization strategy for the model weights.
+    target_decoder_layers : `int`, optional (default = `1`)
```
Comment (JohnGiorgi) [inline allennlp_models/generation/models/copynet_seq2seq.py:71]: I think the `target` in `target_decoder_layers` might be unnecessary? Maybe `num_decoder_layers` is more informative?

Comment (epwalsh) [reply allennlp_models/generation/models/copynet_seq2seq.py:71]: I agree ✋

**Comment by epwalsh** [comment]
@annajung, do you see this option on the left side of the actions output screen?

image

Maybe outside contributions can't re-run jobs. Another way to trigger a re-run would be to push an empty commit:

```
git commit --allow-empty -m "trigger build"
```

**Comment by epwalsh** [comment]
I had to restart the server that runs the GPU tests. Looks fine now.

File: allennlp_models/generation/models/copynet_seq2seq.py (hunk 3 above)
Comment (annajung) [reply allennlp_models/generation/models/copynet_seq2seq.py]: done!

**Comment by annajung** [comment]
yeah, I dont see the retrigger option. 


> git commit --allow-empty -m "trigger build"

good to know for the future, thanks!

**Comment by annajung** [comment]
closing the PR based on https://github.com/allenai/allennlp/issues/4766#issuecomment-815131999 and https://github.com/allenai/allennlp/issues/4766#issuecomment-815143074.

**Pull request closed without merging**